### PR TITLE
vault: support k8s auth method

### PR DESCRIPTION
**Pull request by Kryvchun**
Issue: https://github.com/hashicorp/envconsul/issues/274

This PR implements support of K8S auth method:
1. Add new fields to [CreateConsulClientInput](https://github.com/hashicorp/consul-template/blob/5a20fdd92ecb4dd7de9f1fbb0430827d5ff4cfc9/dependency/client_set.go#L84):
    ```go
    K8SAuthRoleName            string
    K8SServiceAccountMountPath string
    K8SServiceAccountToken     string
    K8SServiceMountPath        string
    ```
2. If a token is set in the function [CreateVaultClient](https://github.com/hashicorp/consul-template/blob/5a20fdd92ecb4dd7de9f1fbb0430827d5ff4cfc9/dependency/client_set.go#L339) then ignore k8s auth method.
3. Execute [client.Auth().Login(...)](https://pkg.go.dev/github.com/hashicorp/vault/api#Auth.Login) with [vault.auth.KubernetesAuth](https://pkg.go.dev/github.com/hashicorp/vault/api/auth/kubernetes#NewKubernetesAuth) auth method,
4. set the token.

This can't be integration tested, so I mocked one endpoint (`/v1/auth/kubernetes/login`) with an `*httptest.Server`, all other calls are proxied to Vault

### CODE REVIEW

**Review by ygworldr — APPROVED** [review]
LGTM

File: config/vault.go [hunk 1]
```diff
@@ -34,6 +34,14 @@ const (
 	// DefaultLeaseRenewalThreshold is the default fraction of a non-renewable
 	// lease to wait for before refreshing
 	DefaultLeaseRenewalThreshold = .90
+
+	// DefaultK8SServiceAccountTokenPath is a default path to a file
+	// with service token for the k8s auth method.
+	DefaultK8SServiceAccountTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
```
Comment (yellowmegaman) [inline config/vault.go:40]: Should be /run/secrets/kubernetes.io/serviceaccount/token

File: config/vault.go [hunk 2]
```diff
@@ -83,6 +91,34 @@ type VaultConfig struct {
...
+	// authentication makes it easy to introduce a Vault token into
+	// a Kubernetes Pod.
+	//
+	// This can also be set via the VAULT_K8S_AUTH_ROLE_NAME.
+	K8SAuthRoleName *string `mapstructure:"k8s_auth_role_name"`
+	// K8SServiceAccountTokenPath is the path of file that contains
+	// a K8SServiceAccountToken. It will be ignored if K8SServiceAccountToken
+	// is set.
+	//
+	// Default value is "/var/run/secrets/kubernetes.io/serviceaccount/token".
```
Comment (yellowmegaman) [inline config/vault.go:109]: Should be /run/secrets/kubernetes.io/serviceaccount/token

File: dependency/client_set.go [hunk 3]
```diff
@@ -521,3 +535,45 @@ func (c *ClientSet) Stop() {
...
+		opts = append(opts, vaultkubernetesauth.WithServiceAccountToken(
+			i.K8SServiceAccountToken,
+		))
+	case i.K8SServiceAccountTokenPath != "":
+		opts = append(opts, vaultkubernetesauth.WithServiceAccountTokenPath(
+			i.K8SServiceAccountTokenPath,
+		))
+	default:
+		// The Kubernetes service account token JWT will be retrieved
+		// from /var/run/secrets/kubernetes.io/serviceaccount/token.
```
Comment (yellowmegaman) [inline dependency/client_set.go:556]: Should be /run/secrets/kubernetes.io/serviceaccount/token

**Review by yellowmegaman — APPROVED** [review]
LGTM
tested with auth role

File: config/vault.go (hunk 1 above)
Comment (eikenb) [reply config/vault.go:40]: Thanks for the review @yellowmegaman!

You say LGTM *and* suggest a change... does it need that change first, and then can be merged or is it good without that change and can be merged as is? Thanks.

File: dependency/client_set.go (hunk 3 above)
Comment (hedhyw) [reply dependency/client_set.go]: @yellowmegaman changed

File: config/vault.go (hunk 2 above)
Comment (hedhyw) [reply config/vault.go]: @yellowmegaman changed

File: config/vault.go (hunk 1 above)
Comment (hedhyw) [reply config/vault.go]: @eikenb @yellowmegaman changed

Comment (Kryvchun) [reply config/vault.go]: I updated it here and in [envconsul]( https://github.com/hashicorp/envconsul/pull/281/commits/67ee7c33f1ce78b7d34f4335fa001a2ed832ce69)

Comment (yellowmegaman) [reply config/vault.go]: @eikenb sorry for confusion, it was "change needed, but otherwise LGTM"

**Comment by Kryvchun** [comment]
@eikenb Hi. Is everything okay with the implementation? Or do I need to fix something?

**Comment by eikenb** [comment]
Hey @Kryvchun, sorry for the silence. I was planning on working on this when I got to working on Envconsul which took a little longer than expected but I did start working on a new release for it this week and I plan on reviewing this as part of that (so I can loop it into the Envconsul release).

**Review by eikenb — APPROVED** [review]
Looks good and checks all the required boxes, thanks @Kryvchun!

Thanks also for the help reviewing and testing @ygworldr, @yellowmegaman!!! Super appreciate that as I have limited time to validate the k8ts side of things.

**Comment by Kryvchun** [comment]
> How to use this with consul-template ? Is there any doc ?

You can configure (inside `vault {}` config):
- `k8s_auth_role_name`: string (role name, it enables this auth method)
- `k8s_service_mount_path`: string (a part of k8s login path, `auth/:value/login`, default: `kubernetes`)

Token:
- `k8s_service_account_token_path`: string (path to secret, default: `/run/secrets/kubernetes.io/serviceaccount/token`)
Or
- `k8s_service_account_token`: string (value of an account token for k8s auth method).

There is no docs for this in `consul-template`, only in `envconsul`.

**Comment by solomon2201** [comment]
Good day. @Kryvchun  and @eikenb I have a problem with your update. Previously we are using VaultClient  in a init_container for take a token from vault and consul-template for took a secret and rotate access to vault. But we need connect to vault always because we are have dynamic env.

I made update to your commit. Its working well, consul take all secret from vault but we are lose token in 1h. I turned on the trace log for debug it. consul-template don`t rotate the token. 
I understand, its just working with variable only ( but for token we need use VaultClient)
#vault_agent_token_file = "/tmp/vault/agent/token"
Can you add rotating token in consul-template for your new option??  We are really need that.


its update block code
```
vault {
  - vault_agent_token_file = "/var/run/secrets/vaultproject.io/.vault-token"  
 + k8s_service_account_token_path = "/var/run/secrets/kubernetes.io/serviceaccount/token"
 + k8s_service_mount_path =  "kubernetes"
 + k8s_auth_role_name = "app"
  unwrap_token           = false
  renew_token            = true
  retry {
    enabled     = true
    attempts    = 10
    attempts    = 3
    backoff     = "250ms"
    max_backoff = "1m"
  }

```
and vault accessor. You can see its not rotating.

```Key                 Value
---                 -----
accessor            bW8hQQQQaKSVxIR
creation_time       1674067534
creation_ttl        1h
display_name        kubernetes-app-app
entity_id           qqqqq-bcaa-55df-43ba-qqqqqqq
expire_time         2023-01-18T19:45:34.181388978Z
explicit_max_ttl    0s
id                  n/a
issue_time          2023-01-18T18:45:34.181398288Z
meta                map[role:app service_account_name:app service_account_namespace:app service_account_secret_name: service_account_uid:qqqq-qqq-490a-qqqqq-76d2863696d1]
num_uses            0
orphan              true
path                auth/kubernetes/login
policies            [app-database-creds-kubernetes app-secrets-rw default]
renewable           true
ttl                 29s
type                service
```


Thanks for your time

**Pull request merged**